jpl.pl

```diff
@@ -4333,6 +4351,209 @@
+exception_msg(ep02,'when 1st arg denotes a class, 2nd arg must be the name of a public, static, non-final field').
+exception_msg(ep03,'class has no public static fields of this name').
+exception_msg(ep04,'cannot assign a value to a final field').
+exception_msg(ep05,'the value is not assignable to the named field of the class').
+exception_msg(ep06,'3rd arg does not represent any Java value or object').
+exception_msg(ep07,'more than one public static field of the class has this name (this should not happen)(?)').
+
+% ---
+
+exception_msg(eq01,'not all values are assignable to the array element type').
+exception_msg(eq02,'not all values are convertible to Java values or references').
+exception_msg(eq03,'array element type is unknown (this should not happen)').
+
+% ---
+
+exception_msg(er01,'must be acyclic').
+
+% ---
+
+exception_msg(es01,'1st arg must be bound to a JPL type').
+
+% The predicates that throw. 
+% Pred+    is generally the predicate descriptor of the predicate raising the exception.
+% MsgCode+ is the atom indicating which cleartext message to insert into the exception term.
+
+from_msg_code(MsgCode,MsgText) :-
+   (exception_msg(MsgCode,MsgText) -> true ; (MsgText = MsgCode)).
+
+throw_instantiation_error(Pred,MsgCode) :-
+   from_msg_code(MsgCode,MsgText),
+   Formal  = instantiation_error,
+   Context = context(Pred,MsgText),
+   throw(error(Formal,Context)).
+
+throw_domain_error(Pred,ValidDomain,Culprit,MsgCode) :-
+   from_msg_code(MsgCode,MsgText),
+   Formal  = domain_error(ValidDomain,Culprit),
+   Context = context(Pred,MsgText),
+   throw(error(Formal,Context)).
+
+throw_type_error(Pred,ValidType,Culprit,MsgCode) :-
+   from_msg_code(MsgCode,MsgText),
+   Formal  = type_error(ValidType,Culprit),
+   Context = context(Pred,MsgText),
+   throw(error(Formal,Context)).
+
+throw_existence_error(Pred,ObjectType,Culprit,MsgCode) :-
+   from_msg_code(MsgCode,MsgText),
+   Formal = existence_error(ObjectType,Culprit),
+   Context = context(Pred,MsgText),
+   throw(error(Formal,Context)).
+
+throw_permission_error(Pred,Operation,PermissionType,Culprit,MsgCode) :-
+   from_msg_code(MsgCode,MsgText),
+   Formal  = permission_error(Operation,PermissionType,Culprit),
+   Context = context(Pred,MsgText),
+   throw(error(Formal,Context)).
+
+% throw_illegal_state_error/2 is not ISO but we need it! (there should
+% be an ISO error like that, really).
+%
+% It is used in place of the original approach of throwing a system_error
+% with parameters. system_errors with parameters is not ISO either --
+% and dubious: system_error should be used when the harddisk crashes etc,
+% not when the assertion fails.
+
+throw_illegal_state_error(Pred,MsgCode) :-
+   from_msg_code(MsgCode,MsgText),
+   Formal  = illegal_state_error,
+   Context = context(Pred,MsgText),
+   throw(error(Formal,Context)).  % TODO maybe not have functor "error" but "jpl_error"?
```

Comment:
This should be error/2.   error/2 are in general errors (as opposed to timeouts, aborts and user specific exceptions that are not errors) and error/2 is recognised by the debugger to take action under some condition.   A name `state_error` would be more in line with the way Prolog names errors.   You could call this `jpl_state_error`.   
I'd be in favor of simply using the Formal and Context as arguments to the throw/1, but that is more style.

Reply:
@JanWielemaker , I just want to clarify that you recommend this NOT to use `error/2`, but to use some new term like `jpl_state_error` and then add the corresponding rule you mention. 

Just asking because your first sentence says: "this should be error/2" but I think that is a typo?

Reply:
No, I got it wrong. We **must** use `error/2` We are referring about the "non-standard" `illegal_state_error` term here. See below.